pytoda/datasets/utils/collator.py

```diff
@@ -0,0 +1,125 @@
+import numpy as np
+from typing import Tuple, List
+import torch
+from torch.utils.data._utils.collate import default_convert
+from pytoda.datasets.utils.factories import BACKGROUND_TENSOR_FACTORY
+
+
+class Collator:
+    """Contains function to pad data returned by dataloader."""
+
+    def __init__(
+        self,
+        padding_mode: List,
+        padding_values: List,
+        max_length: int,
+        batch_first: bool = True,
+        device: torch.device = (
+            torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+        ),
+    ):
+        """Constructor.
+
+        Args:
+            padding_mode (List): List of string values, one for each item in the tuple of a batch, indicating the type of background tensor to generate.
+            padding_values (List): List of values to fill in the background tensor for each item in the tuple of a batch.
+            dim (int): Dimension of the data.
+            max_length (int): Maximum set length.
+            batch_first (bool, optional): Whether batch size is the first
+                dimension or not. Defaults to True.
+            device (torch.device): device where the tensors are stored.
+                Defaults to gpu, if available.
+        """
+        super(Collator, self).__init__()
+        # self.background_tensor = super().constant_value_tensor()
+        self.padding_mode = padding_mode
+        self.padding_values = padding_values
+        # self.dim = dim
+        self.max_len = max_length
+        self.batch_first = batch_first
+        self.device = device
+
+    def __call__(self, DataLoaderBatch: Tuple) -> Tuple:
+        """Collate function for batch-wise padding of samples.
+
+        Args:
+            DataLoaderBatch (Tuple): Tuple of tensors returned by get_item of the
+            dataset class.
+
+        Returns:
+            Tuple: Tuple of padded input tensors and tensor of set lengths.
+        """
+
+        batch_size = len(DataLoaderBatch)
+        # convert [(a,b,c),(e,f,g)] to [[a,e],[b,f],[c,g]]
+        batch_split = list(zip(*DataLoaderBatch))
+
+        returned_tensors = []
+
+        for index, batch in enumerate(batch_split):
+
+            if self.padding_mode == "none":
+                returned_tensors.append(batch)
+            else:
+
+                mode = self.padding_mode[index]
+                value = self.padding_values
+                trailing_dims = batch.size()[2:]
+
+                if self.batch_first:
+                    out_dims = (batch_size, self.max_len, trailing_dims)
+                else:
+                    out_dims = (self.max_len, batch_size, trailing_dims)
+
+                out_tensor = BACKGROUND_TENSOR_FACTORY[mode](
+                    value, out_dims, device=self.device
+                )
+
+                for datum_index, tensor in enumerate(default_convert(batch)):
+
+                    length = tensor.size(0)
```

Comment:
Unlike the suggestion I provided, you split the batch into batched data first, and then loop over the items.
I think that is fine for the input we have in mind (item is a tuple of data that can be converted to tensor <- Needs to be documented) and allows different kinds of padding (including none) to be done with a list of modes as inputs.

We should not use default_convert anymore, as this is intended to convert a single item. It would work here too since batch is a Sequence, but it's rather confusing than just making sure we have a tensor here:
```suggestion
                for batch_index, datum in enumerate(batch):
                    tensor = torch.as_tensor(datum)
                    length = tensor.size(0)
```

DataLoaderBatch should be renamed so something like `item_list`. the input type is `List`(see https://github.com/pytorch/pytorch/blob/671ee71ad4b6f507218d1cad278a8e743780b716/torch/utils/data/dataloader.py#L30) for any collate_fn and `List[Tuple]` for this case, as it does not support datasets returning not a tuple but a single element.

Going this route, I would force passing padding_mode**s** and padding_values as lists with same length as the item-tuples.
Then iterate with a zip
```
for batch, padding_mode, padding_value in zip(batch_split, self.padding_modes, self.padding_values):
```

Reply:
self.padding_mode  should never be "none".
if one does not want to use the collator, don't use it.

iterating `batch_split` not doing padding for some datum could be done with `None` elements in passed padding_mode (list). Take care to document what happens then (must be tensor? will be made tensor? returns list?)

Reply:
I added the "none" for cases where there exists certain tensors in a tuple that do not need padding but other tensors do. so for example if each item is (features, target, lengths), then self.padding_mode = ['constant','range','none']

Reply:
`['constant','range', None]`
strings need to be keys in the factory, `None` is the exception and marked as such by not being a string.

Reply:
`padding_mode: List[Union[str, None]]`